### PR TITLE
Refactor inter thread communication

### PR DESCRIPTION
**Pull request by chme**
Hi @ejurgensen,

as shortly discussed via mail, this pr introduces a generic helper for communicating between threads running a libevent eventloop. Especially the changes in player.c need a thorough review.

### CODE REVIEW

**Comment by ejurgensen** [comment]
I had a look at it now, and basically I think it looks very promising!

There are a lot of changes, so I haven't exactly done a line-by-line-review, and I probably won't get to do that either. I would, however, like to look some more at it. Here are my comments from what I have looked at so far:

1. You recommend a review of the player.c changes - but there are a lot of them :-) Any particular areas you had in mind?
2. I must admit the flow in commands.c is a bit hard to follow. That is not your fault, I know you have based it on the existing. What do you think about this? Do you think it could be simplified?
3. Memory management is also complex. Take for instance commands.c line 77 where cmd is free'd, but cmd->arg can live on (referenced by some other pointer). Not wrong, I think, but not simple either. So I was thinking if you have made any tests with Valgrind or the like?
4. Adding to that, I think I did spot a minor memleak in command_exec_sync. You calloc cmd, but do not free in the "if (ret < 0)". In this case I would think it would be better to use the stack and not calloc at all?
5. I was wondering if it would be possible to also include the thread exit stuff in this? Not in this PR maybe, but later?
6. Noticed a few places where you need to adjust the comments to match the updated code: worker.c line 281 is misplaced now, and worker.c line 94. There might be other places as well, hope you will check up on that.
7. I don't think you need to cast a calloc(). In fact I believe it is recommended not to. Sometimes you haven't done it, but in commands.c you do.

**Comment by chme** [comment]
>  1) You recommend a review of the player.c changes - but there are a lot of them :-) Any particular areas you had in mind?

What player.c makes special is, that some functions (e. g.  setting volume) do return COMMAND_PENDING with pending events > 1. I tried to keep the current behavior, but it would be great if you could check that.

> 2) I must admit the flow in commands.c is a bit hard to follow. That is not your fault, I know you have based it on the existing. What do you think about this? Do you think it could be simplified?

I added a commit and some code comments, but i do not think, that (maybe beside better documentation) the logic could be simplified without losing functionality. There were small differences at the different places (e.g. player.c with the pending events or worker.c with a delayed execution) and i tried to consolidate them in commands.c which makes it a bit more complex. At least the complexity is now not duplicated at various places.

> 3) Memory management is also complex. Take for instance commands.c line 77 where cmd is free'd, but cmd->arg can live on (referenced by some other pointer). Not wrong, I think, but not simple either. So I was thinking if you have made any tests with Valgrind or the like?

I did some tests with Valgrind and did not notice anything obvious (but i am not sure if i hit every change). That cmd->arg is not freed if the async command returns COMMAND_PENDING is only used by worker.c if delay > 0. An alternative would be to let worker.c make another copy of the argument and in command_cb always free the argument. Not sure if this would make it better, what do you think?

> 4) Adding to that, I think I did spot a minor memleak in command_exec_sync. You calloc cmd, but do not free in the "if (ret < 0)". In this case I would think it would be better to use the stack and not calloc at all?

Changed it to use the stack like you suggested.

> 5) I was wondering if it would be possible to also include the thread exit stuff in this? Not in this PR maybe, but later?

Yes, that makes sense. I would prefer to do it in a separate pr.

> 6) Noticed a few places where you need to adjust the comments to match the updated code: worker.c line 281 is misplaced now, and worker.c line 94. There might be other places as well, hope you will check up on that. 

I hope, i found all places. 

> 7) I don't think you need to cast a calloc(). In fact I believe it is recommended not to. Sometimes you haven't done it, but in commands.c you do.

Removed the casts.

**Comment by chme** [comment]
Fyi, I tried adding the exit_cb/thread_exit functionality into the commands util: https://github.com/chme/forked-daapd/tree/threadexit

**Comment by ejurgensen** [comment]
Thanks for this, hereby merged. I had a look at the thread exit, and that also looks good. I think my only comment is concerning the naming, I think it would be nice to have command_base_new and commands_base_free (or _destroy or something like that). The latter would break the loop and clean up the commands_base.
